lib/kube/proxy/portforward_websocket.go

```diff
@@ -0,0 +1,310 @@
+			// Kubernetes API server.
+			// Although users can connect via Websocket, Teleport connection between
+			// its components or Kubernetes API server is done using SPDY client
+			// which requires request_id.
+			requestID: fmt.Sprintf("%d", port),
+		}
+
+		portBytes := make([]byte, 2)
+		binary.LittleEndian.PutUint16(portBytes, port)
+		// Protocol requires sending the port to identify which channels belong to
+		// each port.
+		dataStream.Write(portBytes)
+		errorStream.Write(portBytes)
+	}
+
+	h := &websocketPortforwardHandler{
+		conn:          conn,
+		streamPairs:   streamPairs,
+		podName:       req.podName,
+		targetConn:    targetConn,
+		onPortForward: req.onPortForward,
+		FieldLogger: logrus.WithFields(logrus.Fields{
+			trace.Component:   teleport.Component(teleport.ComponentProxyKube),
+			events.RemoteAddr: req.httpRequest.RemoteAddr,
+		}),
+		context: req.context,
+	}
+	// run the portforward request until termination.
+	h.run()
+	return nil
+}
+
+// extractTargetPortsFromStrings extracts the desired ports from the request
+// query parameters.
+func extractTargetPortsFromStrings(portsStrings []string) ([]uint16, error) {
+	if len(portsStrings) == 0 {
+		return nil, trace.BadParameter("query parameter %q is required", PortHeader)
+	}
+
+	ports := make([]uint16, 0, len(portsStrings))
+	for _, portString := range portsStrings {
+		if len(portString) == 0 {
+			return nil, trace.BadParameter("query parameter %q cannot be empty", PortHeader)
+		}
+		for _, p := range strings.Split(portString, ",") {
+			port, err := parsePortString(p)
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+			ports = append(ports, port)
+		}
+	}
+
+	return ports, nil
+}
+
+// websocketChannelPair represents the error and data streams for a single
+// port.
+type websocketChannelPair struct {
+	port        uint16
+	requestID   string
+	dataStream  io.ReadWriteCloser
+	errorStream io.ReadWriteCloser
+}
+
+// websocketPortforwardHandler is capable of processing a single port forward
+// request over a websocket connection
+type websocketPortforwardHandler struct {
+	conn          *wsstream.Conn
+	streamPairs   []*websocketChannelPair
+	podName       string
+	targetConn    httpstream.Connection
+	onPortForward portForwardCallback
+	logrus.FieldLogger
+	context context.Context
+}
+
+// run ivokes the targetConn SPDY connection and copies the client data into
+// the targetConn and the responses into the targetConn data stream.
+// If any error occurs, stream is closed an the error is sent via errorStream.
+func (h *websocketPortforwardHandler) run() {
+	wg := sync.WaitGroup{}
+	wg.Add(len(h.streamPairs))
+
+	for _, pair := range h.streamPairs {
+		p := pair
+		go func() {
+			defer wg.Done()
+			h.portForward(p)
+		}()
+	}
+
+	wg.Wait()
+}
+
+// portForward copies the client and upstream streams.
+func (h *websocketPortforwardHandler) portForward(p *websocketChannelPair) {
+	defer p.dataStream.Close()
+	defer p.errorStream.Close()
```

Comment:
I placed them there because we use the `errorStream` to send the returned error back to the user.
https://github.com/gravitational/teleport/pull/18791/files/95299546214802e33bd6d1c7b876cfd4482ae0f8#diff-975e7e305dc75c442ad250e946d07435f3fd0bac13e31b1c5a1028781460ccbcR224